### DISCUSSION TITLE
Authenticator configs left to database when removing an authenticator

### DISCUSSION THREAD

**Discussion by akikoskinen** [keycloak/keycloak#17040] [Q&A]
Let's say one has an authentication flow and it contains some configurable authenticators. When such a configured authenticator is removed (via admin UI or REST API) from the authentication flow, its authenticator config is left into the database (tables `authenticator_config` and `authenticator_config_entry`).

Is there a reason why the authenticator configs aren't also removed?

Tried with Keycloak 19.0.2, but have seen this in previous versions already.

**Comment by oquirion** [comment]
When deleting a flow, it cascade and delete it's child executions, BUT if those child are bound to authentication config, then these config are left 'dangling' in the database. We have every realm setup as Infra as Code and this behavior actually cause sort of a db record leak.
After many deployment, the system got slow as hell, causing contention on these 2 tables : AUTHENTICATOR_CONFIG and AUTHENTICATOR_CONFIG_ENTRY.

We add to manually clean up these 2 tables with SQL queries and change our deploy script such as:

`function delete_flow {
  local alias="${1}"
  local flow
  local flow_id
  local executions
  local authenticationConfig_id

  if flow=$(kc_r_get authentication/flows --fields id,alias --format csv --noquotes | grep "$alias"); then
    executions=$(kc_r_get authentication/flows/"$alias"/executions | jq -r .[].authenticationConfig)
    
    for authenticationConfig_id in $executions; do
      if [ "$authenticationConfig_id" != "null" ]; then
          kc_r_delete "authentication/config/$authenticationConfig_id"
          log "Deleted authentication config '$authenticationConfig_id'"
      fi
    done
 
    flow_id=$(echo "$flow" | awk -F, '{ print $1 }')
    kc_r_delete "authentication/flows/$flow_id"
    log "Deleted flow '$flow_id'"
  fi
}`

We now basically remove all configuration when we delete a flow.
I hope this helps.

**Comment by docid00** [comment]
This is still a thing in 24.0.2. I opened an issue for that as i am almost sure that its unintended.
https://github.com/keycloak/keycloak/issues/28428

**Comment by docid00** [comment]
@akikoskinen fixed in https://github.com/keycloak/keycloak/issues/24795
